packages/cli/src/utils/upgrades.ts

```diff
@@ -0,0 +1,72 @@
+import {output, path, file} from '@shopify/cli-kit';
+import {renderInfo, renderSuccess} from '@shopify/cli-kit/node/ui';
+import {applyTransform, type Transform} from './transform.js';
+import {type Changeset} from './parse-guide.js';
+import glob from 'fast-glob';
+
+export interface UpgradeOptions {
+  dry?: boolean;
+  silent?: boolean;
+  diff?: boolean;
+}
+
+export async function runChangesets(
+  directory: string,
+  changes: Changeset[],
+  transforms: Transform[] | Transform,
+  options: UpgradeOptions = {},
+) {
+  const files = await glob(path.join(directory, '/**/*.{js,ts,tsx,jsx}'), {
+    ignore: ['**/node_modules/**', '**/dist/**'],
+  });
+
+  const tasks = changes.map((changeset) => {
+    const {title = '', description = ''} = changeset;
+
+    return {
+      title,
+      task: async () => {
+        const results = await applyTransform(transforms, files);
+        if (!options.silent) output.info(description);
+
+        const changed = Array.from(results.entries())
+          .filter(([_, result]) => result.state === 'changed')
+          .map(([filename, result]) => ({
+            ...result,
+            filename,
+            pathname: path.relative(directory, filename),
+          }));
+
+        for (const {pathname, filename, after, diff} of changed) {
+          if (options.diff) {
+            renderInfo({
+              headline: pathname,
+              body: output.content`${output.token.linesDiff(diff)}`.value,
```

Comment:
In the video the diff that shows is for the whole file, which in some cases will be a lot to dump to the terminal, and (ideally?) these should be pretty granular. Could we truncate that output in some sensible way to put the focus on the actual changes?